buf/registry/plugin/v1beta1/upload_service.proto

```diff
@@ -0,0 +1,64 @@
+// Copyright 2023-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+package buf.registry.plugin.v1beta1;
+
+import "buf/registry/plugin/v1beta1/commit.proto";
+import "buf/registry/plugin/v1beta1/label.proto";
+import "buf/registry/plugin/v1beta1/plugin.proto";
+import "buf/registry/priv/extension/v1beta1/extension.proto";
+import "buf/validate/validate.proto";
+
+option go_package = "buf.build/gen/go/bufbuild/registry/protocolbuffers/go/buf/registry/plugin/v1beta1";
+
+// Upload contents.
+service UploadService {
+  // Upload contents for given set of Plugins.
+  //
+  // Contents are expected to be WASM modules.
+  rpc Upload(UploadRequest) returns (UploadResponse) {}
+}
+
+message UploadRequest {
+  option (buf.registry.priv.extension.v1beta1.message).request_only = true;
+
+  // Content to upload.
+  message Content {
+    // The Plugin of the reference.
+    PluginRef plugin_ref = 1 [(buf.validate.field).required = true];
+    // The content to upload.
+    bytes content = 2 [(buf.validate.field).required = true];
+    // The labels to associate with the Commit for the Content.
+    //
+    // If an id is set, this id must represent a Label that already exists and is
+    // owned by the Plugin. The Label will point to the newly-created Commits for the References,
+    // or will be updated to point to the pre-existing Commit for the Reference.
+    //
+    // If no labels are referenced, the default Label for the Plugin is used.
+    //
+    // If the Labels do not exist, they will be created.
+    // If the Labels were archived, they will be unarchived.
+    repeated ScopedLabelRef scoped_label_refs = 3;
+  }
+  // The Contents of all references.
+  repeated Content contents = 1 [(buf.validate.field).repeated.min_items = 1];
```

Comment:
Plugins might get quite large, so we should consider the batched nature of this API.

Unlike modules, users will likely be pushing a single plugin. Furthermore, I don't think we have plans for the `buf plugin push` command to support multiple plugin uploads.

Lastly, there's a cost in keeping these in memory. That was one of the motivations for compressing the content payload, because we can keep a smaller footprint in memory. E.g., the ml plugin compressed compressed 16MB -> 3MB. We also need to invoke and validate each plugin.

Options:

1. Keep this as repeated, but add a validation rule for exactly one
2. Remove repeated, allow 1 upload at a time

Reply:
Every other API in registry-proto operates on batches. In the absence of a concrete performance problem, there is no current reason to change the API to be inconsistent with other APIs. Therefore, no change here is required at this time.